internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type MarkerWatcher interface {
+	Watch() <-chan UpdateMarker
+	Run(ctx context.Context)
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	}
+
+	// Watch the upgrade marker file's directory, not the file itself, so we
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
```

Comment:
Why do we add the files to watch here but we instantiate the goroutine to consume events only in Run() ?

Can't we instantiate the watcher directly in `Run()` so that we can also add a 
```
defer watcher.Close()
```
on the goroutine, making sure we close the watcher ?
If that is the case maybe we should document it on the interface stating that,  whatever happens,  a created watcher must always be closed, even if we didn't call Run() or other functions on it

Reply:
Yes, it should be possible to do all this work inside `Run()`.  Let me give it a shot.

Reply:
The reason I added the files to watch in the constructor instead of `Run()` was because I only wanted the files to be added to the watch once per `MarkerFileWatcher` instance.  In theory, the calling code could call `Run()` repeatedly on the same `MarkerFileWatcher` instance, and I didn't want to add the same files to the watch each time.  I don't know how `fsnotify` behaves if the same path is watched multiple times — will it emit duplicate events? Or is it smart enough to de-dupe paths each time they're added? 

Let me play with `fsnotify` some more to see if watching the same paths is safe.  If it is, I think we can greatly simplify the implementation here.

Reply:
I experimented with `fsnotify` and confirmed that the `Add()` method is safe.  With that, I was able to simplify the implementation of the `UpgradeMarkerWatcher` quite a bit in 8b9d67c797.